plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,402 @@
+
+				workingBuffer[i + offset][0] *= fadeValue;
+				workingBuffer[i + offset][1] *= fadeValue;
+			}
+		}
+
+		instrumentTrack()->processAudioBuffer(workingBuffer, frames + offset, handle);
+
+		emit isPlaying(noteDone, sliceStart, sliceEnd);
+	}
+	else { emit isPlaying(-1, 0, 0); }
+}
+
+// uses the spectral flux to determine the change in magnitude
+// resources:
+// http://www.iro.umontreal.ca/~pift6080/H09/documents/papers/bello_onset_tutorial.pdf
+void SlicerT::findSlices()
+{
+	if (m_originalSample.frames() <= 1) { return; }
+	m_slicePoints = {};
+
+	const int windowSize = 512;
+	const float minBeatLength = 0.05f; // in seconds, ~ 1/4 length at 220 bpm
+
+	int sampleRate = m_originalSample.sampleRate();
+	int minDist = sampleRate * minBeatLength;
+
+	float maxMag = -1;
+	std::vector<float> singleChannel(m_originalSample.frames(), 0);
+	for (int i = 0; i < m_originalSample.frames(); i++)
+	{
+		singleChannel[i] = (m_originalSample.data()[i][0] + m_originalSample.data()[i][1]) / 2;
+		maxMag = std::max(maxMag, singleChannel[i]);
+	}
+
+	// normalize and find 0 crossings
+	std::vector<int> zeroCrossings;
+	float lastValue = 1;
+	for (int i = 0; i < singleChannel.size(); i++)
+	{
+		singleChannel[i] /= maxMag;
+		if (sign(lastValue) != sign(singleChannel[i]))
+		{
+			zeroCrossings.push_back(i);
+			lastValue = singleChannel[i];
+		}
+	}
+
+	std::vector<float> prevMags(windowSize / 2, 0);
+	std::vector<float> fftIn(windowSize, 0);
+	std::array<fftwf_complex, windowSize> fftOut;
+
+	fftwf_plan fftPlan = fftwf_plan_dft_r2c_1d(windowSize, fftIn.data(), fftOut.data(), FFTW_MEASURE);
+
+	int lastPoint = -minDist - 1; // to always store 0 first
+	float spectralFlux = 0;
+	float prevFlux = 1E-10; // small value, no divison by zero
+	float real, imag, magnitude, diff;
+
+	for (int i = 0; i < singleChannel.size() - windowSize; i += windowSize)
+	{
+		// fft
+		std::copy_n(singleChannel.data() + i, windowSize, fftIn.data());
+		fftwf_execute(fftPlan);
+
+		// calculate spectral flux in regard to last window
+		for (int j = 0; j < windowSize / 2; j++) // only use niquistic frequencies
+		{
+			real = fftOut[j][0];
+			imag = fftOut[j][1];
+			magnitude = std::sqrt(real * real + imag * imag);
+
+			// using L2-norm (euclidean distance)
+			diff = std::sqrt(std::pow(magnitude - prevMags[j], 2));
+			spectralFlux += diff;
+
+			prevMags[j] = magnitude;
+		}
+
+		if (spectralFlux / prevFlux > 1.0f + m_noteThreshold.value() && i - lastPoint > minDist)
+		{
+			m_slicePoints.push_back(i);
+			lastPoint = i;
+			if (m_slicePoints.size() > 128) { break; } // no more keys on the keyboard
+		}
+
+		prevFlux = spectralFlux;
+		spectralFlux = 1E-10; // again for no divison by zero
+	}
+
+	m_slicePoints.push_back(m_originalSample.frames());
+
+	for (float& sliceValue : m_slicePoints)
+	{
+		int closestZeroCrossing = *std::lower_bound(zeroCrossings.begin(), zeroCrossings.end(), sliceValue);
+		if (abs(sliceValue - closestZeroCrossing) < windowSize)
```

Comment:
I believe `std::abs` should be used here.

Reply:
Done